### PR TITLE
Issue 1131 | Drop keycloak

### LINKED ISSUE #1131: Drop keycloak support
### What is the expected enhancement?

Drop keycloak support in gateway.

### PR DESCRIPTION
**Pull request by IceKhan13**
### Summary

Drop keycloak

### Details and comments

Closes #1131

### CODE REVIEW

**Comment by psschwei** [comment]
Before merging, I'd like to test this against local Kubernetes too... (I'll run on kind once the PR is out of draft)

**Comment by Tansito** [comment]
I had almost the same so I'm fine with this PR, @IceKhan13 . What I would do is simplify it a bit, let keycloak stuff and remove everything in another PR, what do you think?

**Comment by Tansito** [comment]
We have tons of configurations related with keycloak that it will be better review in apart from this I think.

**Comment by IceKhan13** [comment]
> We have tons of configurations related with keycloak that it will be better review in apart from this I think.

Do you mean configurations in helm charts? 

> What I would do is simplify it a bit

most of the changes in just notebooks diffs :) 
if you look at code lines it will be like 50 lines of change, which is simple, I would say 😄 

I do not mind if you close this PR and push your changes, I was just sitting in a meeting and decided to take this one.

**Comment by Tansito** [comment]
> Do you mean configurations in helm charts?

Mostly yeah, but around applications too. Something that we can clean in the next PR.

> most of the changes in just notebooks diffs :)
> if you look at code lines it will be like 50 lines of change, which is simple, I would say 😄
>
> I do not mind if you close this PR and push your changes, I was just sitting in a meeting and decided to take this one.

Yeah, no it's fine. Just to mention in case you were thinking to include to remove all the configuration stuff. That maybe we can let it for another PR.

**Comment by Tansito** [comment]
Should we remove this from `client` too, @IceKhan13 ?:

https://github.com/Qiskit-Extensions/quantum-serverless/blob/main/client/quantum_serverless/core/provider.py#L455

**Comment by IceKhan13** [comment]
Oh, let me remove from client to. YOu are right.

**Comment by psschwei** [comment]
Got an error when running on Kind:

`QuantumServerlessException: Cannot verify token.`

(did a `pip install .` on client, so should be using the most recent updates...)

File: docker-compose.yaml [hunk 1]
```diff
@@ -77,19 +54,17 @@ services:
       - SETTING_KEYCLOAK_REALM=quantum-serverless
       - SETTINGS_KEYCLOAK_CLIENT_SECRET=supersecret
```
Comment (Tansito) [inline docker-compose.yaml:55]: Here I think we can just remove `SETTING_KEYCLOAK_*` configuration

**Comment by psschwei** [comment]
> Did you change the authentication mechanism 

No, wasn't expecting that I'd need to (I assumed it would work out of the box)... let me try again

**Comment by psschwei** [comment]
hmmm... still fails with 

```json
 {
    "name": "SETTINGS_AUTH_MECHANISM",
    "value": "mock_token"
  },
```

File: gateway/main/settings.py [hunk 2]
```diff
@@ -207,6 +207,9 @@
 
 DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
 
+# =============
+# AUTH SETTINGS
+# =============
 SETTINGS_AUTH_MECHANISM = os.environ.get("SETTINGS_AUTH_MECHANISM", "default")
 SETTINGS_DEFAULT_AUTH_CLASSES = [
```
[the remark this answers is not in the document]
Comment (IceKhan13) [reply gateway/main/settings.py:214]: We still need them if we want to support open auth eventually, I think

File: docker-compose.yaml (hunk 1 above)
Comment (IceKhan13) [reply docker-compose.yaml]: resolved in https://github.com/Qiskit-Extensions/quantum-serverless/pull/1132/commits/08c62782bba3344fe358a1c31a2149472269dea9

**Comment by IceKhan13** [comment]
@psschwei did you rebuild gateway image?

**Comment by psschwei** [comment]
> @psschwei did you rebuild gateway image?

yes, built from source

**Comment by psschwei** [comment]
envvar is being picked up inside the gateway pod...

```bash
I have no name!@gateway-7b444889b6-dkjh8:/usr/src/app$ python
Python 3.9.16 (main, May 23 2023, 14:24:31) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ.get("SETTINGS_AUTH_MECHANISM")
'mock_token'
>>> 
```

**Comment by IceKhan13** [comment]
🤔

**Comment by psschwei** [comment]
looks like the keycloak pod is still there... let me try deleting that (or maybe I should wait for the helm config update in the next PR...)

**Comment by IceKhan13** [comment]
keycloak chart should not affect it by any means

File: gateway/main/settings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tansito) [reply gateway/main/settings.py:214]: Ok, understood!

**Comment by psschwei** [comment]
does the postgres username matter? looks like I was using `bn_keycloak`, but not sure if that user is gone now...

**Comment by IceKhan13** [comment]
nope, it does not matter

**Comment by Tansito** [comment]
Helm should continue being the same with the difference in the auth.mechanism value I think 🤔

**Comment by psschwei** [comment]
Ok, I figured out the error...  for reference, this was the code block:

```python
serverless = ServerlessProvider(
    token=os.environ.get("GATEWAY_TOKEN", "awesome_token"),
    host=os.environ.get("GATEWAY_HOST", "http://localhost:8010"),
)
serverless
```

Can anyone spot the problem? :smile:

**Comment by Tansito** [comment]
Damn port?

File: gateway/api/authentication.py [hunk 3]
```diff
@@ -90,3 +90,25 @@ def authenticate(self, request):
...
+        token = None
+
+        auth_header = request.META.get("HTTP_AUTHORIZATION")
+        if auth_header is not None:
+            token = auth_header.split(" ")[-1]
+
+            if settings.SETTINGS_AUTH_MOCK_TOKEN is not None:
+                if token == settings.SETTINGS_AUTH_MOCK_TOKEN:
+                    try:
+                        user = User.objects.get(username="mock_user")
```
Comment (psschwei) [inline gateway/api/authentication.py:109]:
```suggestion
                        user = User.objects.get(username="mock-user")
```

Underscore is an invalid character for subdomains, so this causes an error on kubernetes

File: gateway/api/authentication.py [hunk 4]
```diff
@@ -90,3 +90,25 @@ def authenticate(self, request):
...
+                    except User.DoesNotExist:
+                        user = User(username="mock_user")
```
Comment (psschwei) [inline gateway/api/authentication.py:111]:
```suggestion
                        user = User(username="mock-user")
```

File: gateway/tests/api/test_authentication.py [hunk 5]
```diff
@@ -106,3 +106,18 @@ def test_with_nested_verification_fields(self):
...
+
+    def test_mock_auth(self):
+        """Tests for mock authentication backend."""
+        backend = MockAuthBackend()
+        request = MagicMock()
+        request.META.get.return_value = "Bearer my_awesome_token"
+
+        with self.settings(SETTINGS_AUTH_MOCK_TOKEN="my_awesome_token"):
+            user, token = backend.authenticate(request)
+            self.assertEqual(user.username, "mock_user")
```
Comment (psschwei) [inline gateway/tests/api/test_authentication.py:118]:
```suggestion
            self.assertEqual(user.username, "mock-user")
```

**Review by psschwei — CHANGES_REQUESTED** [review]

**Comment by psschwei** [comment]
> Damn port?

You win :laughing:

**Comment by IceKhan13** [comment]
ref: https://github.com/Qiskit-Extensions/quantum-serverless/issues/1133

**Comment by IceKhan13** [comment]
Tested it with rancher locally too. Got the same error like Paul :)

**Review by psschwei — APPROVED** [review]
Worked for me on Kubernetes, code changes look good

**Comment by IceKhan13** [comment]
> Tomorrow I can upload helm configuration.

Perfect!
